File: app/src/api/learning.ts
import type { ApiResponse, PaginatedData } from '@zhiqu/shared';
import type { LearningTask, TaskProgress, SubmitProgressRequest } from '@zhiqu/shared';
import { client, unwrap, unwrapPaged } from './client';

// ── Backend schemas (local mirrors) ──
export interface TaskOut {
  id: string;
  title: string;
  task_type: string;
  status: string;
  created_at: string;
  due_date?: string;
}

export interface TaskDetail extends TaskOut {
  items: unknown[];
}

export interface ProgressOut {
  id: string;
  task_id: string;
  student_id: string;
  status: string;
  score?: number;
  started_at?: string;
  completed_at?: string;
}

export interface MasteryRecordOut {
  id: string;
  student_id: string;
  knowledge_point_id: string;
  mastery_level: number;
  updated_at: string;
}

export interface StudySessionOut {
  id: string;
  student_id: string;
  knowledge_point_id?: string;
  session_type: string;
  started_at: string;
  ended_at?: string;
}

// ── 任务编排（教师分配的任务，学生视角）──

export function listMyTasks(params: { status?: string; page?: number; page_size?: number } = {}) {
  return client.get<ApiResponse<PaginatedData<TaskOut>>>('/app/learning/tasks', { params }).then(unwrapPaged);
}

export function listTasks(params: { status?: string; page?: number; page_size?: number } = {}) {
  return listMyTasks(params);
}

export function getTask(taskId: string) {
  return client.get<ApiResponse<TaskDetail>>(`/app/learning/tasks/${taskId}`).then(unwrap);
}

// TODO: 后端尚无独立的 task-items 端点；TaskDetail.items 已含子项，此处返回 task.items
export async function listTaskItems(taskId: string) {
  const task = await getTask(taskId);
  return (task as TaskDetail).items ?? [];
}

export function startTask(taskId: string) {
  return client.post<ApiResponse<ProgressOut>>(`/app/learning/tasks/${taskId}/start`).then(unwrap);
}

export function submitTask(taskId: string, data: SubmitProgressRequest) {
  return client.post<ApiResponse<ProgressOut>>(`/app/learning/tasks/${taskId}/submit`, data).then(unwrap);
}

// ── 学习任务（LearningCoreService，学生自主学习）──

export function listLearningTasks(params: { status?: string; page?: number; page_size?: number } = {}) {
  return client.get<ApiResponse<PaginatedData<LearningTask>>>('/app/learning/learning-tasks', { params }).then(unwrapPaged);
}

export function submitLearningTask(taskId: string, data: { score?: number; answer_snapshot?: unknown }) {
  return client.post<ApiResponse<LearningTask>>(`/app/learning/learning-tasks/${taskId}/submit`, data).then(unwrap);
}

// ── 掌握度 ──

export function listMastery(params: { page?: number; page_size?: number } = {}) {
  return client.get<ApiResponse<PaginatedData<MasteryRecordOut>>>('/app/learning/mastery', { params }).then(unwrapPaged);
}

// ── 学习会话 ──

export function createStudySession(data: { knowledge_point_id?: string; session_type?: string }) {
  return client.post<ApiResponse<StudySessionOut>>('/app/learning/study-sessions', data).then(unwrap);
}

export function listStudySessions(params: { page?: number; page_size?: number } = {}) {
  return client.get<ApiResponse<PaginatedData<StudySessionOut>>>('/app/learning/study-sessions', { params }).then(unwrapPaged);
}

// ── 学习概览 ──

export function getMyProgress() {
  return client.get<ApiResponse<unknown>>('/app/learning/progress').then(unwrap);
}

File: services/interfaces/api/app/learning.py
"""App 学习路由 — 任务 / 进度 / 学习会话 / 掌握度"""

from __future__ import annotations

from uuid import UUID
from typing import Optional

from fastapi import APIRouter, Query

from shared.response import ok, paged
from interfaces.schemas.learning import (
    TaskOut, TaskDetail,
    ProgressStart, ProgressSubmit, ProgressOut,
    LearningTaskOut, LearningTaskCreate, LearningTaskUpdate,
    MasteryRecordOut,
    StudySessionCreate, StudySessionOut,
)
from interfaces.api.deps import CurrentUser, StudentUser, LearningSvc, LearningCoreSvc

router = APIRouter(prefix="/api/v1/app/learning", tags=["app-learning"])


# ── 任务编排（学生视角）──────────────────────────────────

@router.get("/tasks", summary="我的任务列表")
async def my_tasks(
    user: CurrentUser,
    svc: LearningSvc,
    status: Optional[str] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    items, total = await svc.list_tasks_for_student(
        student_id=user.sub, status=status,
        page=page, page_size=page_size,
    )
    data = [TaskOut.model_validate(t).model_dump() for t in items]
    return paged(data, total=total, page=page, page_size=page_size)


@router.get("/tasks/{task_id}", summary="任务详情")
async def get_task(task_id: UUID, user: CurrentUser, svc: LearningSvc):
    task = await svc.get_task(task_id)
    return ok(TaskDetail.model_validate(task).model_dump())


@router.post("/tasks/{task_id}/start", summary="开始任务")
async def start_task(task_id: UUID, user: CurrentUser, svc: LearningSvc):
    progress = await svc.start_progress(task_id, student_id=user.sub)
    return ok(ProgressOut.model_validate(progress).model_dump())


@router.post("/tasks/{task_id}/submit", summary="提交任务进度")
async def submit_task(
    task_id: UUID, body: ProgressSubmit,
    user: CurrentUser, svc: LearningSvc,
):
    progress = await svc.submit_progress(task_id, student_id=user.sub, data=body)
    return ok(ProgressOut.model_validate(progress).model_dump())


# ── 学习任务（原 learning_core，学生端）─────────────────

@router.get("/learning-tasks", summary="我的学习任务")
async def my_learning_tasks(
    user: CurrentUser,
    svc: LearningCoreSvc,
    status: Optional[str] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    items, total = await svc.list_tasks(
        student_id=user.sub, status=status,
        page=page, page_size=page_size,
    )
    data = [LearningTaskOut.model_validate(t).model_dump() for t in items]
    return paged(data, total=total, page=page, page_size=page_size)


@router.post("/learning-tasks/{task_id}/submit", summary="提交学习任务答案")
async def submit_learning_task(
    task_id: UUID, body: LearningTaskUpdate, user: CurrentUser, svc: LearningCoreSvc,
):
    result = await svc.submit_task(
        task_id,
        score=body.score or 0,
        answer_snapshot=body.answer_snapshot,
    )
    return ok(LearningTaskOut.model_validate(result).model_dump())


# ── 掌握度 ────────────────────────────────────────────

@router.get("/mastery", summary="我的知识点掌握度")
async def my_mastery(
    user: CurrentUser,
    svc: LearningCoreSvc,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    items, total = await svc.list_mastery(
        student_id=user.sub, page=page, page_size=page_size,
    )
    data = [MasteryRecordOut.model_validate(m).model_dump() for m in items]
    return paged(data, total=total, page=page, page_size=page_size)


# ── 学习会话 ──────────────────────────────────────────

@router.post("/study-sessions", summary="记录学习会话")
async def create_study_session(
    body: StudySessionCreate, user: CurrentUser, svc: LearningCoreSvc,
):
    session = await svc.create_study_session(
        student_id=user.sub,
        knowledge_point_id=body.knowledge_point_id,
        session_type=body.session_type,
    )
    return ok(StudySessionOut.model_validate(session).model_dump())


@router.get("/study-sessions", summary="学习会话列表")
async def list_study_sessions(
    user: CurrentUser,
    svc: LearningCoreSvc,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    items, total = await svc.list_study_sessions(
        student_id=user.sub, page=page, page_size=page_size,
    )
    data = [StudySessionOut.model_validate(s).model_dump() for s in items]
    return paged(data, total=total, page=page, page_size=page_size)


# ── 学习概览 ──────────────────────────────────────────

@router.get("/progress", summary="学习进度概览")
async def my_progress(user: CurrentUser, svc: LearningCoreSvc):
    stats = await svc.get_student_progress(student_id=user.sub)
    return ok(stats)
